src/Xrefcheck/Core.hs

```diff
@@ -70,15 +68,89 @@ instance Given ColorMode => Buildable Position where
 
 -- | Full info about a reference.
 data Reference = Reference
-  { rName   :: Text
+  { rName     :: Text
     -- ^ Text displayed as reference.
-  , rLink   :: Text
+  , rLink     :: Text
     -- ^ File or site reference points to.
-  , rAnchor :: Maybe Text
+  , rAnchor   :: Maybe Text
     -- ^ Section or custom anchor tag.
-  , rPos    :: Position
+  , rPos      :: Position
+    -- ^ More info about the link.
+  , rInfo     :: ReferenceInfo
   } deriving stock (Show, Generic)
 
+-- | Info about the reference.
+data ReferenceInfo
+  = RIExternal
+    -- ^ Reference to a file at outer site, e.g @[d](http://www.google.com/doodles)@
+  | RIOtherProtocol
+    -- ^ Entry not to be processed, e.g. @mailto:e-mail@
+  | RIFileLocal
+    -- ^ Reference to this file, e.g. @[a](#header)@
+  | RIFileAbsolute
+    -- ^ Reference to a file absolute to the root, e.g. @[c](/folder/file#header)@
+  | RIFileRelative
+    -- ^ Reference to a file relative to given one, e.g. @[b](folder/file#header)@
+  deriving stock (Show, Generic)
+
+referenceInfo :: Text -> ReferenceInfo
+referenceInfo link = case toString link of
+  []                                    -> RIFileLocal
+  c : _             | isPathSeparator c -> RIFileAbsolute
+  '.' : c : _       | isPathSeparator c -> RIFileRelative
+  '.' : '.' : c : _ | isPathSeparator c -> RIFileRelative
+  _ | hasUrlProtocol                    -> RIExternal
+    | hasProtocol                       -> RIOtherProtocol
+    | otherwise                         -> RIFileRelative
+  where
+    hasUrlProtocol = "://" `T.isInfixOf` T.take 10 link
+    hasProtocol    = ":"   `T.isInfixOf` T.take 10 link
+
+newtype CanonicalPath = UnsafeCanonicalPath
+  { unCanonicalPath :: FilePath
+  } deriving newtype (Show, Eq, Ord)
+
+canonicalizePath :: FilePath -> IO CanonicalPath
+canonicalizePath = (UnsafeCanonicalPath . expandIndirections <$>) . Directory.canonicalizePath
+  where
```

Comment:
`System.Directory.canonicalizePath` does not handle indirections, including "." and "..", when the file does not exist, so I kept this code to expand them anyways in order to not modify the program behaviour. Otherwise, some files that pointed outside the repo will be reported as "File does not exist" instead of "File is outside the repo".

Reply:
Sounds good, I believe we do not effectively call `expandIndirections` more often than before.